### PR TITLE
[Backport release-24.05] go: Backport go 1.23 to nixos-24.05

### PR DESCRIPTION
**Pull request by mglazer**
## Description of changes

go 1.23 is the latest go version.

1.21 has already lost support: https://go.dev/doc/devel/release, this at least gives us some level of ability to take 1.23 and ensure our containers are built with the same go version.

Cherry-picked from: cf4f89ec3f6e7f45ca0e3a2f9f45493088952868

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by eclairevoyant** [comment]
Please follow the guidelines here for backporting:
https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#how-to-backport-pull-requests

particularly cherry-picking, the branch to target, and ensuring that stable doesn't have breakage before backporting

**Comment by winterqt** [comment]
Please cherry-pick "properly," instead of adding your own commit message/comments/etc.

**Comment by eclairevoyant** [comment]
> One thing, there is no single commit that I can pick to cherry pick.

Then you may need to cherry-pick multiple commits. That's fine, but creating an entirely new commit on the release-* branch isn't.

**Comment by mglazer** [comment]
OK, backported everything except for the merge commits.

There was a single merge conflict with: [d82203f](https://github.com/NixOS/nixpkgs/pull/339654/commits/d82203fbe2908e0642779a044454144dd0a91848) which I resolved the best I could.

**Comment by mglazer** [comment]
Is this the right way? It just seemed questionable to backport changes which bring in _other_ unrelated changes.

File: pkgs/development/compilers/go/1.23.nix [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    };
+  };
+
+  meta = with lib; {
+    changelog = "https://go.dev/doc/devel/release#go${lib.versions.majorMinor finalAttrs.version}";
+    description = "Go Programming language";
+    homepage = "https://go.dev/";
+    license = licenses.bsd3;
+    maintainers = teams.golang.members;
+    platforms = platforms.darwin ++ platforms.linux ++ platforms.wasi ++ platforms.freebsd;
```
Comment (mglazer) [inline pkgs/development/compilers/go/1.23.nix:186]: The reason this is failing the cherrypick, is `platforms.wasi` was added here:

https://github.com/NixOS/nixpkgs/commit/050df1a7e6391db584d5bb88da6a845c26f19375

which is before go1.23 existed. I opted to _not_ add additional functionality that wasn't part of the strict set of cherrypicks when resolving the merge conflict. Happy to add platforms.wasi to go1.22 if that's desired.

**Comment by Scrumplex** [comment]
Do we have a chance to get this backported without rebuilding so many packages? https://github.com/NixOS/nixpkgs/pull/344149 needs a backport for a critical vulnerability and upstream updated to Go 1.23.

**Comment by mweinelt** [comment]
Ideally we will not backport the whole version bump, but just the patch for traefik.

**Pull request closed without merging**